### ISSUE TITLE
FatalExecutionEngineError, System.ExecutionEngineException, System.InvalidOperationException + more when running 2 sites in local IIS or IIS Express on my local PC where one site is a virtual site.

### ISSUE THREAD

**Issue by THammond9** [dotnet/efcore#15042]
I have two project that are deployed to the same app pool. One site is the parent and the other is the child (virtual site). We have other legacy ORMs in the parent site. When I navigate the parent site only everything works ok. If I restart the app pool or stop debugging the app and go to the child site everything is ok. If I navigate the parent site using only legacy ORM code and then navigate to the child site both sites are ok. If I navigate the parent site including EF Core and then navigate the chile site the exceptions below are thrown. If I navigate the child site and then the parent site including EF Core the exceptions below are thrown.

tl;dr: I have been able to deduce that EF Core seems to be the cause of the issue when accessed from both sites.

This issue happens on all members of my team so it does not seem to be a local issue. Other members of my team use their own PCs for work so it would not be related to domain policies. I have done research on this issue but have been unable to determine a solution and all other information I have been able to find don't seem to fit this scenario.

This issue happens when using local IIS or IIS Express when debugging and not debugging.

Oddity: This issue does not happen in our Dev or QA environments. Feature is not available in prod but suspect the issue would not occur.

There are 3 exceptions that are associated with this problem. I am pretty sure the last exception is a symptom of the problem. When the error happens I usually get the first two exceptions and then all other methods using EF Core will throw the other exception afterwards. I included it in case it was helpful.

```
Stack trace:
Managed Debugging Assistant 'FatalExecutionEngineError' 
  Message=Managed Debugging Assistant 'FatalExecutionEngineError' : 
'The runtime has encountered a fatal error. 
The address of the error was at 0x7334229e, on thread 0xc34. 
The error code is 0xc0000005. 
This error may be a bug in the CLR or in the unsafe or non-verifiable portions of user code. 
Common sources of this bug include user marshaling errors for COM-interop or PInvoke, 
which may corrupt the stack.'

------------------Next Exception------------------

System.ExecutionEngineException
  HResult=0x80131506
  Message=Exception of type 'System.ExecutionEngineException' was thrown.

------------------Next Exception------------------

System.InvalidOperationException: 
An error occurred when trying to create a controller of type 'MyCompanyApp.Web.Controllers.SignatureController'. 
Make sure that the controller has a parameterless public constructor. ---> System.TypeInitializationException: 
The type initializer for 'Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache' 
threw an exception. ---> 
System.IO.FileLoadException: Loading this assembly would produce a different 
grant set from other instances. (Exception from HRESULT: 0x80131401)
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache..ctor()
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache..cctor()
   --- End of inner exception stack trace ---
   at Microsoft.EntityFrameworkCore.Internal.ServiceProviderCache.get_Instance()
   at Microsoft.EntityFrameworkCore.DbContext..ctor(DbContextOptions options)
   at MyCompanyApp.Entities.Tenant.Tables.ContextTenant..ctor(DbContextOptions options)
   at MyCompanyApp.Entities.Tenant.Context.TenantContext..ctor(DbContextOptions options)
   at MyCompanyApp.Entities.Tenant.Context.TenantContextFactory.CreateDbContext(String[] args)
   at MyCompanyApp.Web.NinjectWebCommon.<>c.<CreateKernel>b__3_1(IContext x) in C:\Users\thomas\source\repos\MyCompanyApp\Simplified\App_Start\NinjectWebCommon.cs:line 48
   at Ninject.Activation.Providers.CallbackProvider`1.CreateInstance(IContext context)
   at Ninject.Activation.Provider`1.Create(IContext context)
   at Ninject.Activation.Context.ResolveInternal(Object scope)
   at Ninject.Activation.Context.Resolve()
   at Ninject.KernelBase.Resolve(IRequest request, Boolean handleMissingBindings)
   at Ninject.KernelBase.Resolve(IRequest request)
   at Ninject.Planning.Targets.Target`1.ResolveWithin(IContext parent)
   at Ninject.Activation.Providers.StandardProvider.GetValue(IContext context, ITarget target)
   at Ninject.Activation.Providers.StandardProvider.<>c__DisplayClass15_0.<Create>b__0(ITarget target)
   at System.Linq.Enumerable.WhereSelectArrayIterator`2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Ninject.Activation.Providers.StandardProvider.Create(IContext context)
   at Ninject.Activation.Context.ResolveInternal(Object scope)
   at Ninject.Activation.Context.Resolve()
   at Ninject.KernelBase.Resolve(IRequest request, Boolean handleMissingBindings)
   at Ninject.KernelBase.Resolve(IRequest request)
   at Ninject.Web.Mvc.NinjectDependencyResolver.GetService(Type serviceType)
   at System.Web.Mvc.DefaultControllerFactory.DefaultControllerActivator.Create(RequestContext requestContext, Type controllerType)
   --- End of inner exception stack trace ---
   at System.Web.Mvc.DefaultControllerFactory.DefaultControllerActivator.Create(RequestContext requestContext, Type controllerType)
   at System.Web.Mvc.DefaultControllerFactory.GetControllerInstance(RequestContext requestContext, Type controllerType)
   at System.Web.Mvc.DefaultControllerFactory.CreateController(RequestContext requestContext, String controllerName)
   at System.Web.Mvc.MvcHandler.ProcessRequestInit(HttpContextBase httpContext, IController& controller, IControllerFactory& factory)
   at System.Web.Mvc.MvcHandler.BeginProcessRequest(HttpContextBase httpContext, AsyncCallback callback, Object state)
   at System.Web.Mvc.MvcHandler.BeginProcessRequest(HttpContext httpContext, AsyncCallback callback, Object state)
   at System.Web.Mvc.MvcHandler.System.Web.IHttpAsyncHandler.BeginProcessRequest(HttpContext context, AsyncCallback cb, Object extraData)
   at System.Web.HttpApplication.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
   at System.Web.HttpApplication.<>c__DisplayClass285_0.<ExecuteStepImpl>b__0()
   at System.Web.HttpApplication.StepInvoker.Invoke(Action executionStep)
   at System.Web.HttpApplication.StepInvoker.<>c__DisplayClass4_0.<Invoke>b__0()
   at Microsoft.AspNet.TelemetryCorrelation.TelemetryCorrelationHttpModule.OnExecuteRequestStep(HttpContextBase context, Action step)
   at System.Web.HttpApplication.<>c__DisplayClass284_0.<OnExecuteRequestStep>b__0(Action nextStepAction)
   at System.Web.HttpApplication.StepInvoker.Invoke(Action executionStep)
   at System.Web.HttpApplication.ExecuteStepImpl(IExecutionStep step)
   at System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously)|MyCompanyApp.DataServices.Services.LogService.Exception|33
```

### Steps to reproduce
Have not tried to repro with a basic app but my setup would be:
1. Create 2 .Net framework 4.6.1 MVC 5 web apps
2. Create .Net Standard 2.0 project for EF Core
3. Change sites' project web configuration to set up a parent > child relationship. ex: Site1-http://localhost:68555 and Site2-http://localhost:68555/child
4. Add Ninject for both sites
5. Use IDesignTimeDbContextFactory to inject an instance of the context
6. Navigate to a method that uses EF Core in both apps without restarting the app pool or stopping the project in Visual Studio.

### Further technical details
EF Core version: 2.2.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio 2017 15.9.9
MVC Version: 5
Site Project .Net Framework: 4.6.1
EF Core Project Framework: Standard 2.0
Using Ninject for DI.
Using IDesignTimeDbContextFactory to create the context.

**Comment by ajcvickers** [comment]
@THammond9 Can you post what "a method that uses EF Core in both apps" looks like? Specifically, how is the context instance created and disposed?

**Comment by THammond9** [comment]
Ninject creating the instance, this is in both app startup methods:
```
kernel.Bind<TenantContext>().ToMethod(x => new TenantContextFactory().CreateDbContext(new[]
{ConfigurationManager.ConnectionStrings[EnvironmentConsts.DbConnectionName].ConnectionString}))
.InRequestScope();
```

The factory creation method, same method for both apps:
```
public TenantContext CreateDbContext(string[] args)
        {
            if (args.Length == 1)
            {
                return new TenantContext(new DbContextOptionsBuilder().UseSqlServer(args.First()).Options);
            }

            throw new InvalidOperationException("An incorrect number of connection string were specified");
        }
```

DI into the constructor of the class. TenantContext is a private get-only property:
```
public DashboardService(TenantContext tenantContext)
        {
            TenantContext = tenantContext;
        }
```

There is no dispose as it is my understanding that Ninject handles the disposing of instances that it is responsible for creating.

I am just using the context like normal. `this.TenantContext.MyTable.Where(x => x.SomeColumn == someValue)` or `(from m in this.TenantContext where m.SomeColumn == someValue select new { m.SomeOtherColumn})` or more complex versions where required (joins, left joins, etc)

**Comment by ajcvickers** [comment]
@AndyAyersMS @briansull Any ideas on this? Notes:
* It's unlikely that EF itself is the root cause given we're just managed code with no ref-emit, although we do compile expressions and execute them.
* I haven't moved to https://github.com/dotnet/coreclr/ because the apps are running on .NET Framework using a .NET Standard class library. Where should .NET Framework issues go?

**Comment by ajcvickers** [comment]
@danmosemsft @karelz @RussKeldorph Any ideas for help or where to route on this?

**Comment by AndyAyersMS** [comment]
Per [dotnet/core's readme](https://github.com/dotnet/core) you should report this through [VS Developer Community Portal](https://developercommunity.visualstudio.com/) for "full" .NET Framework feedback (or via the [Report a Problem tool](https://aka.ms/vs-rap)).

Verifying that the basic app setup as above hits this problem would be useful. 

If that's not possible, then getting a dump file taken at the point of the access violation would probably be enough to make some headway.

**Comment by THammond9** [comment]
I opened an issue in the VS Developer Community Portal. [Here is the link](https://developercommunity.visualstudio.com/content/problem/503347/fatalexecutionengineerror-systemexecutionengineexc.html)

**Comment by ajcvickers** [comment]
Thanks @AndyAyersMS for the info and @THammond9 for filing the new issue.
